### PR TITLE
MTCNN - a Joint Face Detection and Alignment using Multi-task Cascaded Convolutional Networks using TensorFlow framework

### PR DESCRIPTION
**Pull request by look4pritam**
This work is used to reproduce MTCNN, a Joint Face Detection and Alignment using Multi-task Cascaded Convolutional Networks using TensorFlow framework.

1) See <ROOT>/data/WIDER_Face/README.md for downloading WIDER Face dataset.
2) See <ROOT>//data/LFW_Landmark/README.md for downloading LFW facial landmark dataset.

3) Generate a basic dataset i.e. PNet dataset.
python generate_simple_dataset.py \
	--annotation_image_dir=./data/WIDER_Face/WIDER_train/images \ 
	--annotation_file_name=./data/WIDER_Face/WIDER_train/wider_face_train.txt \
	--landmark_image_dir=./data/LFW_Landmark \
	--landmark_file_name=./data/LFW_Landmark/trainImageList.txt \
	--target_root_dir=./data/datasets/mtcnn \
	--minimum_face=12

4) Train PNet.
python train_model.py \
	--network_name=PNet \ 
	--train_root_dir=./data/models/mtcnn/train \
	--dataset_root_dir=./data/datasets/mtcnn \
	--base_learning_rate=0.01 \
	--max_number_of_epoch=30

5) Generate a hard dataset i.e. RNet dataset.
python generate_hard_dataset.py \
	--network_name=RNet \ 
	--train_root_dir=./data/models/mtcnn/train \
	--annotation_image_dir=./data/WIDER_Face/WIDER_train/images \ 
	--annotation_file_name=./data/WIDER_Face/WIDER_train/wider_face_train.txt \
	--landmark_image_dir=./data/LFW_Landmark \
	--landmark_file_name=./data/LFW_Landmark/trainImageList.txt \
	--target_root_dir=./data/datasets/mtcnn \
	--minimum_face=24

6) Train RNet.
python train_model.py \
	--network_name=RNet \ 
	--train_root_dir=./data/models/mtcnn/train \
	--dataset_root_dir=./data/datasets/mtcnn \
	--base_learning_rate=0.01 \
	--max_number_of_epoch=22

7) Generate a hard dataset i.e. ONet dataset.
python generate_hard_dataset.py \
	--network_name=ONet \ 
	--train_root_dir=./data/models/mtcnn/train \
	--annotation_image_dir=./data/WIDER_Face/WIDER_train/images \ 
	--annotation_file_name=./data/WIDER_Face/WIDER_train/wider_face_train.txt \
	--landmark_image_dir=./data/LFW_Landmark \
	--landmark_file_name=./data/LFW_Landmark/trainImageList.txt \
	--target_root_dir=./data/datasets/mtcnn \
	--minimum_face=24

8) Train ONet.
python train_model.py \
	--network_name=ONet \ 
	--train_root_dir=./data/models/mtcnn/train \
	--dataset_root_dir=./data/datasets/mtcnn \
	--base_learning_rate=0.01 \
	--max_number_of_epoch=22

9) Webcamera demo.
python webcamera_demo.py 
OR
python webcamera_demo.py \
	 --test_mode

### CODE REVIEW

**Comment by karmel** [comment]
Thanks, @look4pritam . Typically, models in the Research repository require a collaborator at Google Research to take ownership. Are you working with someone there, or do you have somebody who is willing to act as reviewer and owner of the code?

**Comment by look4pritam** [comment]
I am not working with Google research, but I am ready to take ownership of MTCNN code for next 1-2 years.

**Comment by robieta** [comment]
Would you be so kind as to remove the data and checkpoint files, and replace them with download scripts? (and preprocessing scripts if necessary.) Right now it's hard to determine how substantial this PR is, and those files won't be able to live in the repo regardless.

**Comment by look4pritam** [comment]
Give me some time, I will remove model files and provide download script
for it.

Importance of MTCNN
MTCNN is originally implemented in Caffe.
Face detection  is fundamental step for face recognition.
MTCNN is light weight and does online face detection.
MTCNN combined with InceptionResnet model provides good results for face
recognition with more than 99℅ accuracy on MS-1M dataset with 20K
identities.

Replacing dataset files will take little bit more time.

**Comment by karmel** [comment]
This looks like interesting work, @look4pritam . However, the Research models here are restricted to those affiliated with Google Research for the time being. I would suggest taking a look at [TFHub](https://www.tensorflow.org/hub/), which is designed to allow researchers like yourself to publish new models that are of general interest to the community.

**Comment by look4pritam** [comment]
Thanks for reply. I will upload this project as stand alone project, which will be utilized by community.
